tuftool/src/source.rs

```diff
@@ -8,169 +8,51 @@
-            description: Some(key_id_hex.to_owned()),
-            key_id: key_id.as_ref().cloned(),
-            overwrite: Some(true),
-            type_: "SecureString".to_owned(),
-            value: value.to_owned(),
-            ..rusoto_ssm::PutParameterRequest::default()
-        });
-        tokio::runtime::Runtime::new()
-            .unwrap()
-            .block_on(fut)
-            .context(error::SsmPutParameter {
-                profile: profile.clone(),
-                parameter_name,
-            })?;
-        Ok(())
-    }
-}
-
-impl FromStr for KeySource {
-    type Err = Error;
-
-    #[allow(clippy::find_map)]
-    fn from_str(s: &str) -> Result<Self> {
-        let pwd_url = Url::from_directory_path(std::env::current_dir().context(error::CurrentDir)?)
-            .expect("expected current directory to be absolute");
-        let url = Url::options()
-            .base_url(Some(&pwd_url))
-            .parse(s)
-            .context(error::UrlParse { url: s })?;
-
-        match url.scheme() {
-            "file" => Ok(KeySource::Local(PathBuf::from(url.path()))),
-            #[cfg(any(feature = "rusoto-native-tls", feature = "rusoto-rustls"))]
-            "aws-ssm" => Ok(KeySource::Ssm {
-                profile: url.host_str().and_then(|s| {
-                    if s.is_empty() {
-                        None
-                    } else {
-                        Some(s.to_owned())
-                    }
-                }),
-                parameter_name: url.path().to_owned(),
-                key_id: url
-                    .query_pairs()
-                    .find(|(k, _)| k == "kms-key-id")
-                    .map(|(_, v)| v.into_owned()),
+/// Parses a user-specified source of signing keys.
+/// Currently supported key sources are local files and AWS SSM.
+/// Sources are passed to `tuftool` as arguments in string format:
+/// "file:///" or "aws-ssm:///"
+///
+/// Users are welcome to add their own sources of keys by implementing
+/// the `KeySource` trait in the `tough` library. A user can then add
+/// to this parser to support them in `tuftool`.
+pub(crate) fn parse_key_source(input: &str) -> Result<Box<dyn KeySource>> {
+    let pwd_url = Url::from_directory_path(std::env::current_dir().context(error::CurrentDir)?)
+        .expect("expected current directory to be absolute");
+    let url = Url::options()
+        .base_url(Some(&pwd_url))
+        .parse(input)
+        .context(error::UrlParse { url: input })?;
+
+    match url.scheme() {
+        "file" => Ok(Box::new(LocalKeySource {
+            path: PathBuf::from(url.path()),
+        })),
+        #[cfg(any(feature = "rusoto-native-tls", feature = "rusoto-rustls"))]
+        "aws-ssm" => Ok(Box::new(SsmKeySource {
+            profile: url.host_str().and_then(|s| {
+                if s.is_empty() {
+                    None
+                } else {
+                    Some(s.to_owned())
+                }
```

Comment:
I'm happy to be told otherwise by the Rust experts, but I believe match arms are pretty idiomatic for Rust.

Given that this function performs actions that are fairly specific to `tuftool` (getting a `KeySource` from text), I'm also not convinced that it's something that belongs in the trait.  (I also come from a very OO background so I can absolutely see it your way as well 😄  )